pkg/loadbalancer/loadbalancer.go

```diff
@@ -475,18 +475,16 @@ func (a *L3n4Addr) DeepCopy() *L3n4Addr {
 	}
 }
 
-// Hash calculates a unique string of the L3n4Addr e.g for use as a key in maps
+// Hash calculates a unique string of the L3n4Addr e.g for use as a key in maps.
+// Note: the resulting string meant to be used as a key for maps and is not
+// readable by a human eye when printed out.
 func (a L3n4Addr) Hash() string {
-	const lenIPv4 = 15
 	const lenProto = 1
 	const lenPort = 6
 	const lenScope = 2
 
-	// Note: This capacity will not be enough for long IPv6 addresses, but it
-	// is cheaper to reallocate on overflow than to check the length of a.IP
-	b := make([]byte, 0, lenIPv4+lenProto+lenPort+lenScope)
-
-	b = append(b, a.IP.String()...)
+	b := make([]byte, net.IPv6len, net.IPv6len+lenProto+lenPort+lenScope)
```

Comment:
Not sure whether it really matters but maybe add `+3` for the 3 `b = append(b, '|')` to avoid reallocations in the case where all fields are maxed out?

Reply:
Actually, I figured the separator don't even make sense anymore as the key is anyway unreadable/not meant to be read. So I pruned the 3 `b = append(b, '|')` and it seems to make a little difference for the latest bench case:
```
BenchmarkL3n4Addr_Hash_IPv4-32                  16958146                82.51 ns/op
BenchmarkL3n4Addr_Hash_IPv4_4bytes-32           13032306                99.75 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Short-32            23337658                82.10 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Long-32             13969000                83.09 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Max-32              12060272                98.75 ns/op
```

Reply:
Nice, even better!

Reply:
> I figured the separator don't even make sense anymore as the key is anyway unreadable/not meant to be read.

@rolinh Before this patch, _without separator_, `127.0.0.1:22` and `127.0.0.12:2` would have the same Hash (all other things being equal) right? In other words, the separator is more about isolation than readability.

This PR make it so that the IP part length is constant, which remove the need a separator IP|port separator. As for the port|scope separator, I think it is safe to remove as well since the scope is always either 0 or 1.
https://github.com/cilium/cilium/blob/440c896d10b29e110bb4cc3e708b1fc90c5b616a/pkg/loadbalancer/loadbalancer.go#L170-L175
With that said, I would really like to see a comment on the Scope* const declaration to prevent a new scope like 10 to be added (which without separator would leave us open to Hash collisions, e.g. between `ip=127.0.0.1,port=11,scope=0` and `ip=127.0.0.1,port=1,scope=10`).

Reply:
@kaworu Very good points! You actually gave me an idea which I think solves this issue in an elegant way and improves performance even more.

As a scope is a `uint8` and that [a byte is an alias for `uint8`](https://golang.org/ref/spec#Numeric_types), it can be directly assigned to the byte slice without using `strconv` but a simple cast.
```go
b[net.IPv6len+lenProto] = byte(a.Scope)
```

Similarly, as port is a `uint16`, which is 2 bytes, we can assign each byte individually:
```go
b[net.IPv6len+lenProto+lenScope] = byte(a.Port >> 8)
b[net.IPv6len+lenProto+lenScope+1] = byte(a.Port & 0xff)
```

This means that the byte slice is of constant size so we can preallocate it to the correct size once and then assign each byte individually. I believe this approach solves your concerns and leads to even better results:
```
BenchmarkL3n4Addr_Hash_IPv4-32                  26500321                50.62 ns/op
BenchmarkL3n4Addr_Hash_IPv4_4bytes-32           21352604                57.11 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Short-32            23257256                50.77 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Long-32             24479526                51.06 ns/op
BenchmarkL3n4Addr_Hash_IPv6_Max-32              23630030                50.47 ns/op
```